tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -5002,4 +4991,45 @@ private[optimizer] object OptimizerCore {
       val attemptedInlining: List[AbstractMethodID], cause: Throwable
   ) extends Exception(exceptionMsg(myself, attemptedInlining, cause), cause)
 
+  final class FreshNameAllocator private (
+      private var usedNamesToNextCounter: Map[String, Int]) {
+    import FreshNameAllocator._
+
+    def this() = this(FreshNameAllocator.InitialMap)
+
+    def clear(): Unit = usedNamesToNextCounter = InitialMap
+
+    def freshName(base: String): String = {
+      if (!usedNamesToNextCounter.contains(base)) {
+        usedNamesToNextCounter = usedNamesToNextCounter.updated(base, 1)
+        base
+      } else {
+        var i = usedNamesToNextCounter(base)
+        var result = base + "$" + i
+        while (usedNamesToNextCounter.contains(result)) {
+          i += 1
+          result = base + "$" + i
+        }
+        usedNamesToNextCounter =
+          usedNamesToNextCounter.updated(base, i + 1).updated(result, 1)
```

Comment:
`freshName` is very often called with a non-constant base, coming from the input IR. The archetypal example is [in `withNewLocalDef`](https://github.com/sjrd/scala-js/blob/84fc283c005e5fcfaf774b90f466edf7b4f75739/tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala#L3992). So we cannot rule out `$`s here, unfortunately.